staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -143,6 +143,67 @@ func EachListItem(obj runtime.Object, fn func(runtime.Object) error) error {
 		if takeAddr {
 			raw = raw.Addr()
 		}
+		// raw must be a pointer
+		// allocate a pointer is cheap
+		switch item := raw.Interface().(type) {
+		case *runtime.RawExtension:
+			if err := fn(item.Object); err != nil {
+				return err
+			}
+		case runtime.Object:
+			if err := fn(item); err != nil {
+				return err
+			}
+		default:
+			obj, ok := item.(runtime.Object)
+			if !ok {
+				return fmt.Errorf("%v: item[%v]: Expected object, got %#v(%s)", obj, i, raw.Interface(), raw.Kind())
+			}
+			if err := fn(obj); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// EachListItemWithAlloc invokes fn on each runtime.Object in the list.
+// the difference from  EachListItem is that an object of the same type is regenerated and assigned a value,
+// so that the returned Object list and obj have independent life cycles.
+func EachListItemWithAlloc(obj runtime.Object, fn func(runtime.Object) error) error {
+	if unstructured, ok := obj.(runtime.Unstructured); ok {
+		return unstructured.EachListItem(fn)
+	}
+	// TODO: Change to an interface call?
+	itemsPtr, err := GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	items, err := conversion.EnforcePtr(itemsPtr)
+	if err != nil {
+		return err
+	}
+	len := items.Len()
+	if len == 0 {
+		return nil
+	}
+	takeAddr := false
+	if elemType := items.Type().Elem(); elemType.Kind() != reflect.Pointer && elemType.Kind() != reflect.Interface {
+		if !items.Index(0).CanAddr() {
+			return fmt.Errorf("unable to take address of items in %T for EachListItem", obj)
+		}
+		takeAddr = true
+	}
+
+	for i := 0; i < len; i++ {
+		raw := items.Index(i)
+		if takeAddr {
+			// shallow copy to avoid retaining a reference to the original list item
```

Comment:
Is this the only difference? Is there a reason not to just modify the existing function? This should be very cheap?

Reply:
The reason for not modifying existing functions is that not all scenarios require a shallow copy.

Reply:
Thanks for the suggestion, added a helper function to avoid duplication.